huaweicloud/services/dsc/resource_huaweicloud_dsc_instance.go

```diff
@@ -0,0 +1,437 @@
+	if err != nil {
+		return diag.Errorf("error creating BSS Client: %s", err)
+	}
+
+	payOrderPath := payOrderClient.Endpoint + payOrderHttpUrl
+
+	payOrderOpt := golangsdk.RequestOpts{
+		KeepResponseBody: true,
+		OkCodes: []int{
+			204,
+		},
+	}
+	payOrderOpt.JSONBody = utils.RemoveNil(buildPayOrderBodyParams(orderId.(string)))
+	_, err = payOrderClient.Request("POST", payOrderPath, &payOrderOpt)
+	if err != nil {
+		return diag.Errorf("error pay order=%s: %s", d.Id(), err)
+	}
+
+	bssClient, err := config.BssV2Client(config.GetRegion(d))
+	if err != nil {
+		return diag.Errorf("error creating BSS v2 client: %s", err)
+	}
+	err = common.WaitOrderComplete(ctx, bssClient, orderId.(string), d.Timeout(schema.TimeoutCreate))
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	resourceId, err := common.WaitOrderResourceComplete(ctx, bssClient, orderId.(string), d.Timeout(schema.TimeoutCreate))
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	d.SetId(resourceId)
+	return resourceDscInstanceRead(ctx, d, meta)
+}
+
+func buildCreateDscInstanceBodyParams(d *schema.ResourceData, config *config.Config) map[string]interface{} {
+	bodyParams := map[string]interface{}{
+		"regionId":         config.GetRegion(d),
+		"cloudServiceType": cloudServiceType,
+		"periodNum":        utils.ValueIngoreEmpty(d.Get("period")),
+		"productInfos":     buildCreateDscInstanceRequestBodyProductInfos(d),
+	}
+
+	chargingMode := d.Get("charging_mode").(string)
+	if chargingMode == "prePaid" {
+		bodyParams["chargingMode"] = 0
+	}
+
+	periodUnit := d.Get("period_unit").(string)
+	if periodUnit == "month" {
+		bodyParams["periodType"] = 2
+	} else {
+		bodyParams["periodType"] = 3
+	}
+
+	autoRenew := d.Get("auto_renew").(string)
+	if autoRenew == "true" {
+		bodyParams["isAutoRenew"] = 1
+	} else {
+		bodyParams["isAutoRenew"] = 0
+	}
+
+	return bodyParams
+}
+
+func buildCreateDscInstanceRequestBodyProductInfos(d *schema.ResourceData) []map[string]interface{} {
+	rst := make([]map[string]interface{}, 0, 3)
+	edition := d.Get("edition").(string)
+
+	if edition == resourceSpecCodeStandardBase {
+		rst = append(rst, map[string]interface{}{
+			"cloudServiceType": cloudServiceType,
+			"productId":        "OFFI571544772268847116",
+			"resourceType":     resourceTypeBase,
+			"resourceSpecCode": resourceSpecCodeStandardBase,
+		})
+
+		if size, ok := d.GetOk("obs_expansion_package"); ok {
+			rst = append(rst, map[string]interface{}{
+				"cloudServiceType":      cloudServiceType,
+				"productId":             "OFFI571545095458668555",
+				"resourceType":          resourceTypeObs,
+				"resourceSpecCode":      resourceSpecCodeStandardObs,
+				"resourceSize":          utils.ValueIngoreEmpty(size),
+				"resourceSizeMeasureId": 47,
+			})
+		}
+
+		if size, ok := d.GetOk("database_expansion_package"); ok {
+			rst = append(rst, map[string]interface{}{
+				"cloudServiceType":      cloudServiceType,
+				"productId":             "OFFI571544962480533516",
```

Comment:
**OFFI571544962480533516** is magic string, you can define it with a more meaningful variable.

Reply:
done